infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,314 @@
+package org.corfudb.infrastructure.log.statetransfer;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Builder.Default;
+import lombok.EqualsAndHashCode;
+import lombok.EqualsAndHashCode.Exclude;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.result.Result;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.infrastructure.log.statetransfer.batch.Batch;
+import org.corfudb.infrastructure.log.statetransfer.batch.BatchResult;
+import org.corfudb.infrastructure.log.statetransfer.batchprocessor.StateTransferBatchProcessor;
+import org.corfudb.infrastructure.log.statetransfer.streamprocessor.TransferSegmentFailure;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.NOT_TRANSFERRED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.RESTORED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.TRANSFERRED;
+
+/**
+ * A class responsible for managing a state transfer on the current node.
+ * It executes the state transfer for each non-transferred segment synchronously.
+ */
+@Slf4j
+@AllArgsConstructor
+public class StateTransferManager {
```

Comment:
Why is this only instantiated by the RestoreRedundancyMergeSegments action?if it manages state transfer shouldn't it be used by any of the3 workflows that require state transfer?

Reply:
RestoreRedundancyMergeSegments action is the action that performs a state transfer and also restores the layout and merges segments (that's how it was before my changes too). 
It is the last action in all of the 3 workflows (Look at their constructors).